bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -176,31 +176,39 @@ std::unique_ptr<ClientReplyMsg> ClientsManager::allocateNewReplyMsgAndWriteToSto
                   << KVLOG(c.repliesInfo.size(), maxNumOfReqsPerClient_, clientId, requestSeqNum, replyLength));
   }
 
-  c.repliesInfo.insert_or_assign(requestSeqNum, getMonotonicTime());
   LOG_DEBUG(CL_MNGR, KVLOG(clientId, requestSeqNum));
-  auto r = std::make_unique<ClientReplyMsg>(myId_, requestSeqNum, reply, replyLength);
+  auto r = std::make_shared<ClientReplyMsg>(myId_, requestSeqNum, reply, replyLength);
+  c.repliesInfo.insert({requestSeqNum, r});
   const uint32_t firstPageId = clientIdx * reservedPagesPerClient_;
   LOG_DEBUG(CL_MNGR, "firstPageId=" << firstPageId);
-  uint32_t numOfPages = r->size() / sizeOfReservedPage_;
-  uint32_t sizeLastPage = sizeOfReservedPage_;
-  if (numOfPages > reservedPagesPerClient_) {
-    LOG_FATAL(CL_MNGR,
-              "Client reply is larger than reservedPagesPerClient_ allows"
-                  << KVLOG(clientId, requestSeqNum, reservedPagesPerClient_ * sizeOfReservedPage_, replyLength));
-    ConcordAssert(false);
-  }
+  uint16_t requestNum = 0;
+  for (auto& rep : c.repliesInfo) {
```

Comment:
Do we really need to re-write all replies on reserved pages?

Reply:
As we talked about for now we are going to write all of them each time but we will optimize it to delete the oldest reply from the reserved pages and write the new reply instead.